lib/spack/spack/compilers/__init__.py

```diff
@@ -722,7 +722,9 @@ def _default_make_compilers(cmp_id, paths):
         compiler_cls = spack.compilers.class_for_compiler_name(compiler_name)
         spec = spack.spec.CompilerSpec(compiler_cls.name, version)
         paths = [paths.get(x, None) for x in ("cc", "cxx", "f77", "fc")]
-        target = archspec.cpu.host()
+        # in order for tests to function, this needs to refer to current host
+        # as determined by Spack, not archspec
+        target = archspec.cpu.TARGETS[str(spack.platforms.host().target("default_target"))]
```

Comment:
If this is only needed for tests, then we should probably add a context manager/fixture to overwrite this rather than changing the behavior everywhere (i.e. inside and outside of the tests). What goes wrong when this isn't set?

Reply:
We do access this in the same way that you added elsewhere in the file. However, I worry this could exclude microarchitecture information from the compiler definitions. I have to think about whether that's actually a problem and if so how to resolve that alongside the consistency issue (I do agree there is an inconsistency here)

Reply:
Additional notes:

* This was already generalizing by taking `target.family` below, so there is a generalization
* However, `default_target` is defined on a per-platform basis and can differ from `cpu.host()` outside of testing environments: on Cray, it is the backend target (whereas `cpu.host()` is always the frontend target); this might create a problem if you then index `archspec.cpu.TARGETS`
  * Sub-note: on the Cray system I just checked, these match, but the point is they don't have to (either for Cray, or for any other Platform defined in the future)
* Given the second point, it may be suitable not to index `archspec.cpu.TARGETS`, and to just set `target = str(spack.platforms.host().target("default_target"))`. These are different objects, but we refer to compilers using both `spack.target.Target` and `archspec.cpu.microarchitecture.Microarchitecture` so this seems OK to do

@massimiliano as the person who wrote this line originally, does that seem workable to you to make this code change? Specifically replace

```
target = archspec.cpu.host()
```

with

```
target = spack.platforms.host().target("default_target")
```

? (this would also require discarding the reference to `family` below, so in fact this can be more specific)

Given the above notes I think it does.

Reply:
Right, at the moment, Spack's mocked platform for testing sets an `x86_64` arch, while actually running the tests, archspec picks up the actual system arch, which for me at least is AMD64 (I've had a few others test the code to similar effect on Windows). This results it a number of tests failing with "no compiler for arch" errors. 
In general, I'm surprised this doesn't raise an issue on other platforms that the test suite runs on, so I must be missing something here. I was also a bit confused why there's a potential point for disparities between what Spack understands the platform to be, and what we're determining compilers can target. 

My intent was not to change granularity (maybe a side effect of my unfamiliarity with the archspec module at large) but to shore up a potential vector for internal inconsistency within Spack, and I just see the failing tests as a presentation of that issue, rather than the only place it will ever happen/the explicit cause.

Reply:
> In general, I'm surprised this doesn't raise an issue on other platforms that the test suite runs on, so I must be missing something here. 

It might be a Windows peculiarity to return `AMD64` instead of `x86_64`. For other platforms like macOS we have some basic translation going on in `archspec`. See e.g. https://github.com/archspec/archspec/blob/ed5340947574b1174f70b63edb2848da69e9993f/archspec/cpu/detect.py#L103-L106 Did we consider porting `archspec` to Windows? Would that be a good way forward if it doesn't require a lot of work?

Also, for tests we can `monkeypatch` the `archspec.cpu` module to our liking as a fast solution.